src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,402 @@
+            {
+                resolvedEncoding = GetEncoding(encodingPreference);
+            }
+            else if ( encoding != FileEncoding.Unspecified )
+            {
+                resolvedEncoding = GetEncoding(encoding);
+            }
+            return resolvedEncoding;
+        }
+
+        // This is the way the encoding is implemented in PowerShell 5 and earlier.
+        // If the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
+            {
+                { "microsoft.powershell.commands.addcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportclixmlcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.exportcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.exportpssessioncommand", Encoding.UTF8 }, // with BOM
+                { "microsoft.powershell.commands.formathex", Encoding.ASCII },
+                { "microsoft.powershell.commands.newmodulemanifestcommand", Encoding.Unicode },
+                { "microsoft.powershell.commands.getcontentcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.importcsvcommand", Encoding.ASCII },
+                { "microsoft.powershell.commands.outfilecommand", Encoding.Unicode }, // This includes redirection
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+            };
+
+        /// Get the Windows legacy encoding from our encoding map
+        internal static Encoding GetWindowsLegacyEncoding(Cmdlet cmdlet)
+        {
+            Encoding encoding = Encoding.Default;
+            if ( cmdlet != null )
+            {
+                legacyEncodingMap.TryGetValue(cmdlet.GetType().FullName, out encoding);
+            }
+            return encoding;
+        }
+
+
+        // [System.Text.Encoding]::GetEncodings() | ? { $_.GetEncoding().GetPreamble() } |
+        //     Add-Member ScriptProperty Preamble { $this.GetEncoding().GetPreamble() -join "-" } -PassThru |
+        //     Format-Table -Auto
+        internal static Dictionary<String, FileEncoding> encodingMap =
+            new Dictionary<string, FileEncoding>()
+            {
+                { "255-254", FileEncoding.Unicode },
+                { "254-255", FileEncoding.BigEndianUnicode },
+                { "255-254-0-0", FileEncoding.UTF32 },
+                { "0-0-254-255", FileEncoding.BigEndianUTF32 },
+                { "239-187-191", FileEncoding.UTF8BOM },
+            };
+
+        internal static char[] nonPrintableCharacters = {
```

Comment:
Strictly speaking, `11` (vertical tab) and `12` (form feed) are also printable characters; given their rarity, I assume they were left out intentionally; if so, perhaps a comment would appease drive-by pedants like me.